text/0000-stack-buildpacks.md

```diff
@@ -0,0 +1,599 @@
+
+* A `CNB_STACK_TYPE` env var that a stack buildpack can use to behave differently on each part of the stack
+* Support for `creator`. Because of the new extend phase, it is not possible to easily run the entire buildpack process with a single binary.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Stack buildpacks cannot be provided by the end user (the app developer), they can only by provided by stack creators
+
+# Alternatives
+[alternatives]: #alternatives
+
+- [Add root buildpack as interface for app image extension RFC](https://github.com/buildpacks/rfcs/pull/77)
+- [App Image Extensions (OS packages)](https://github.com/buildpacks/rfcs/pull/23)
+- [Root Buildpacks](https://github.com/buildpacks/rfcs/blob/root-buildpacks/text/0000-root-buildpacks.md): Allow application developers to use root buildpacks in `project.toml`. This would have significant performance implications, and creates a "foot gun" in which end users could build images they are not able to maintain. For this reason, we are holding off on a generic root buildpack feature.
+
+# Prior Art
+[prior-art]: #prior-art
+
+- The term "non-idempotent" is used in section 9.1.2 of [Hypertext Transfer Protocol -- HTTP/1.1 (RFC 2616)](https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html)
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- what happens if a mixin that already exist on the image is required, how does the lifecycle know?
+- what about the bill of materials?
+- what metadata do we need to do a rebase; what are the schema of the labels?
+- how do we prevent problems on rebase if the mixins provided by the stack change?
+- while the rebase process may get its stackpacks from any image passed to it, do we need to support official ways of producing well-formed run-builders?
+
+# Spec. Changes (OPTIONAL)
+[spec-changes]: #spec-changes
+
+A number of changes to the Platform Specification will be required to execute Stack Buildpacks. Those changes will be defined in a separate RFC.
+
+## Stack buildpacks
+
+Stack buildpacks are identical to other buildpacks, with the following exceptions:
+
+1. The `<layers>` directory WILL NOT be used to create layers.
+1. The working directory WILL NOT contain application source code during the build phase.
+1. All changes made to the filesystem during the execution of the stackpack's `bin/build` will be snapshotted and stored as a single layer, with the exception of the following directories:
+
+* `/tmp`
+* `/cnb`
+* `/layers`
+* `/workspace`
+* `/dev`
+* `/sys`
+* `/proc`
+* `/var/run/secrets`
+* `/etc/hostname`, `/etc/hosts`, `/etc/mtab`, `/etc/resolv.conf`
+* `/.dockerenv`
+
+## stack-layer.toml (TOML)
+
+```
+[[excludes]]
+paths = ["<sub-path glob>"]
+cache = false
+restore = false
+```
+
+Where:
+
+* `paths` = a list of paths to exclude from the layer
+* `cache` = if true, the paths will be excluded from the launch image layer, but will be included in the cache layer.
```

Comment:
> excluded from the launch image layer
Only makes sense for layers created during the `extend` phase. For layers created during the `build` phase we should remove these paths from the filesystem before executing any regular buildpacks.

Reply:
addressed in b9cf2b1